fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java

```diff
@@ -409,4 +410,11 @@ protected void setOptimizeForIndexFilters(boolean shouldOptimizeForIndexFilters)
         public T value;
     }
 
+    protected static void assertMatches(@Nonnull final RecordQueryPlan plan, @Nonnull BindingMatcher<? extends RecordQueryPlan> planMatcher) {
+        assertTrue(planMatcher.matches(plan));
```

Comment:
Can at least add `, plan::getString` or something like that to give _some_ message.

Reply:
Will do!

Reply:
Here is an example of the new explain functionality. Interestingly, it is now also possible to explain why a matcher didn't match an object, but for some matchers the wording would be extremely heavy as it would have to list out all the matches that didn't match for let's say an `exactlyInAnyOrder()`. Didn't want to do that for this PR but we can add that in the future.

```
Index(MySimpleRecord$num_value_3_indexed [EQUALS $__in_num_value_3_indexed__0]) WHERE __in_num_value_3_indexed__0 IN [1, 2, 4]
 does not match exactly
plan match { 
    case recordQueryInValuesJoinPlan if plan instanceOf[RecordQueryInValuesJoinPlan] => 
        all of {
            recordQueryInValuesJoinPlan match { 
                case planChildren(collection) => 
                    all recordQueryIndexPlan in collection {
                        recordQueryIndexPlan match { 
                            case recordQueryIndexPlan if recordQueryIndexPlan instanceOf[RecordQueryIndexPlan] => 
                                all of {
                                    recordQueryIndexPlan match { 
                                        case recordQueryIndexPlan => 
                                            all of {
                                                recordQueryIndexPlan match { 
                                                    case quantifiers(collection) => 
                                                        case collection if collection isEmpty() => success
                                                } && 
                                                recordQueryIndexPlan match { 
                                                    case indexName(object) => 
                                                        match object { case MySimpleRecord$num_value_3_indexed => success }
                                                }
                                            }
                                    } && 
                                    recordQueryIndexPlan match { 
                                        case comparisons(scanComparisons) => 
                                            scanComparisons match { 
                                                case range(object) => 
                                                    match object { case [EQUALS $__in_num_value_3_indexed__0] => success }
                                            }
                                    }
                                }
                        }
                    }
            } && 
            recordQueryInValuesJoinPlan match { 
                case values(object) => 
                    match object { case [1, 2, 4] => success }
            }
        }
}
```
Note that I adopted a scala-like syntax which heavily influenced writing these matchers in the first place.